proof/refine/ARM/Reply_R.thy

```diff
@@ -1081,27 +1024,841 @@ lemma sc_replies_relation_replyNext_update:
+lemma sc_replies_relation_scReply_cross:
+  "\<lbrakk> sc_replies_relation s s'; pspace_relation (kheap s) (ksPSpace s');
+     pspace_aligned s; pspace_distinct s; valid_objs s;
+   \<exists>n. kheap s ptr = Some (kernel_object.SchedContext sc n)\<rbrakk>
+    \<Longrightarrow> scReplies_of s' ptr = hd_opt (sc_replies sc)"
+  apply (prop_tac "sc_at' ptr s'")
+   apply (fastforce intro!: sc_at_cross valid_sched_context_size_objsI simp: obj_at_def is_sc_obj_def)
+  apply (clarsimp simp: sc_replies_relation_def obj_at'_def projectKOs)
+  apply (drule_tac x=ptr and y="sc_replies sc" in spec2)
+  apply (prop_tac "sc_replies_of s ptr = Some (sc_replies sc)")
+   apply (clarsimp simp: vs_heap_simps)
+  apply (clarsimp simp: sc_of_def opt_map_def)
+  apply (case_tac "sc_replies sc"; clarsimp simp: projectKO_opt_sc)
+  done
+
+(** sc_with_reply and sc_replies_relations : crossing information **)
+
+lemma sc_replies_relation_prevs_list':
+  "\<lbrakk> sc_replies_relation s s';
+     kheap s scp = Some (kernel_object.SchedContext sc n)\<rbrakk>
+    \<Longrightarrow> heap_ls (replyPrevs_of s') (scReplies_of s' scp) (sc_replies sc)"
+  apply (clarsimp simp: sc_replies_relation_def sc_replies_of_scs_def scs_of_kh_def map_project_def)
+  apply (clarsimp simp: opt_map_left_Some sc_of_def)
+  done
+
+(* FIXME RT: move to TcbAcc_R? also, drop alignment conditions from TcbAcc_R.pspace_relation_tcb_at *)
+lemma pspace_relation_sc_at:
+  assumes p: "pspace_relation (kheap s) (ksPSpace s')"
+  assumes t: "scs_of' s' scp \<noteq> None"
+  shows "sc_at scp s" using assms
+  by (fastforce elim!: pspace_dom_relatedE obj_relation_cutsE
+                 simp: other_obj_relation_def is_sc_obj obj_at_def projectKOs
+                split: Structures_A.kernel_object.split_asm if_split_asm
+                       ARM_A.arch_kernel_obj.split_asm)
+
+lemma sc_replies_relation_sc_with_reply_cross_eq_pred:
+  "\<lbrakk> sc_replies_relation s s'; pspace_relation (kheap s) (ksPSpace s')\<rbrakk>
+    \<Longrightarrow> (\<exists>sc n.
+          kheap s scp = Some (kernel_object.SchedContext sc n) \<and>
+          rp \<in> set (sc_replies sc)) =
+      (\<exists>xs. heap_ls (replyPrevs_of s') (scReplies_of s' scp) xs
+             \<and> rp \<in> set xs)"
+  apply (rule iffI; clarsimp)
+   apply (fastforce dest: sc_replies_relation_prevs_list' heap_path_takeWhile_lookup_next)
+  apply (case_tac "scReplies_of s' scp", simp)
+  apply (rename_tac p)
+  apply (drule pspace_relation_sc_at[where scp=scp])
+   apply (clarsimp simp: projectKOs opt_map_left_Some)
+  apply (clarsimp simp: obj_at_def is_sc_obj)
+  apply (drule (1) sc_replies_relation_prevs_list', clarsimp simp: opt_map_left_Some)
+  apply (drule (1) heap_ls_unique, clarsimp)
+  done
+
+(* crossing equality for sc_with_reply *)
+lemma sc_replies_relation_sc_with_reply_cross_eq:
+  "\<lbrakk> sc_replies_relation s s'; pspace_relation (kheap s) (ksPSpace s')\<rbrakk>
```

Comment:
nitpick: extra spaces + indentation (if extra space at start, do we expect extra space at end?)